docs/src/main/asciidoc/application-data-caching-guide.adoc

```diff
@@ -0,0 +1,433 @@
+
+=== @CacheResult
+
+Loads a method result from the cache without executing the method body whenever possible.
+
+When a method annotated with `@CacheResult` is invoked, Quarkus will compute a cache key from the method arguments (or generate a key if the method has no arguments)
+and use it to check in the cache whether the method has been already invoked.
+The key computation is done from all the method arguments if none of them is annotated with `@CacheKey`, or all the arguments annotated with `@CacheKey` otherwise.
+If a value is found in the cache, it is returned and the annotated method is never actually executed.
+If no value is found, the annotated method is invoked and the returned value is stored in the cache using the computed or generated key.
+
+A method annotated with `CacheResult` is protected by a lock on cache miss mechanism.
+If several threads try to retrieve a cache value from the same missing key concurrently, the method will only be invoked once.
+The first calling thread will trigger the method invocation while the subsequent calling threads will wait for the end of the invocation to get the cached result.
+The `lockTimeout` parameter can be used to interrupt the lock after a given delay.
+See the parameter Javadoc for more details.
+
+This annotation can neither be used on a method returning `void` nor in combination with another Quarkus method caching annotation.
+
+=== @CacheInvalidate
+
+Removes an entry from the cache.
+
+When a method annotated with `@CacheInvalidate` is invoked, Quarkus will compute a cache key from the method arguments (or generate a key if the method has no arguments)
+and use it to try to remove an existing entry from the cache.
+The key computation is done from all the method arguments if none of them is annotated with `@CacheKey`, or all the arguments annotated with `@CacheKey` otherwise.
+If the key does not identify any cache entry, nothing will happen.
+
+This annotation can't be used in combination with another Quarkus method caching annotation.
+
+[TIP]
+====
+If the `@CacheResult` or `@CacheInvalidate` annotations are used on a method with no parameters, a unique default cache key derived from the cache name will be generated and used.
+====
+
+=== @CacheInvalidateAll
+
+When a method annotated with `@CacheInvalidateAll` is invoked, Quarkus will remove all entries from the cache.
+
+This annotation can't be used in combination with another Quarkus method caching annotation.
+
+=== @CacheKey
+
+When a method argument is annotated with `@CacheKey`, it is identified as a part of the cache key during an invocation of a
+method annotated with `@CacheResult` or `@CacheInvalidate`.
+
+This annotation is optional and should only be used when some of the method arguments are NOT part of the cache key.
+
+== Configuring the underlying caching provider
+
+For now, https://github.com/ben-manes/caffeine[Caffeine] is the only available caching provider in Quarkus.
+We plan to add more providers soon.
```

Comment:
I don't think that should be part of the doc and I don't think that is true.

Reply:
Well, there were [discussions about other caching providers](https://github.com/quarkusio/quarkus/issues/3128#issuecomment-511351463) when this all started 6 months ago. Maybe what was written at that time about introducing Infinispan as another caching provider is no longer true now. Anyway, I can update the doc to avoid mentioning future caching providers.

Reply:
I changed the extension guide.